src/charm.py

```diff
@@ -157,6 +199,17 @@ def process_unremoved_units(self, event) -> None:
                 logger.error("Deferring process_unremoved_units: error=%r", e)
                 event.defer()
 
+    def _on_mongodb_relation_joined(self, event: ops.charm.RelationJoinedEvent) -> None:
+        """Add peer to replica set.
+
+        Args:
+            event: The triggering relation joined event.
+        """
+        self._on_mongodb_relation_handler(event)
+
+        if self.unit.is_leader():
+            self.update_app_relation_data()
+
```

Comment:
```suggestion
        # allow leader to update relation data and hosts if it isn't leaving
        if not self.unit.is_leader():
            return 
       
       self._on_mongodb_relation_handler(event)
        try:
            self.update_app_relation_data()
        except PyMongoError as e:
            logger.error("Deferring on updating app relation data since: error: %r", e)
            event.defer()
            return

```